README.md
# Sage Pay Suite integration for Magento 2

[CHANGELOG](https://github.com/ebizmarts/magento2-sage-pay-suite/blob/master/CHANGELOG.md)

## Installation Instructions

## Composer installation (preferred method)
1. Run this command on Magento's root dir.

`composer config repositories.ebizmarts composer https://packages.ebizmarts.com`

2. Run this command on Magento's root dir. Contact us to get your access token.

`composer config http-basic.sagepaysuite.gitlab.ebizmarts.com token your_token`

3. Run this command to get the module.

`composer require ebizmarts/sagepaysuite`

4. Install the module.

`bin/magento setup:upgrade`

##Token expired and it's causing problems.
If your token expired and it's causing problems when you run composer update. 
You can renew your support to get a new token or simply run `composer config --unset repositories.ebizmarts`

## Manual installation

__Requirements__

  - The ZIP file named **Ebizmarts_SagePaySuiteM2-1.1.22.zip**
  - Access to the Magento 2 server via SSH
  - The unzip command should be available, check by running `which unzip`
  - [Composer](https://getcomposer.org/) needs to be installed in the server

__Installation__
 
  1. Upload the ZIP file to the Magento 2 server.

  2. Get access to the Magento 2 server.

  3. Go to the Magento2 modules folder.

    $ cd $MAGENTO_FOLDER$/app/code

  4. Create the directory (if it does not exist) that will hold the module contents
    `$ mkdir -p Ebizmarts/SagePaySuite`
   
  5. Go to the SagePaySuite folder
  
    $ cd $MAGENTO_FOLDER$/app/code/Ebizmarts/SagePaySuite
   
  6. Uncompress Sage Pay Suite package
  
    $ unzip /PATH/TO/PACKAGE/Ebizmarts_SagePaySuiteM2-1.1.22.zip

  7. This will create the following content in $MAGENTO_FOLDER$/app/code
    <pre>
    └── Ebizmarts
        └── SagePaySuite
            ├── Api
            ├── Block
            ├── Controller
            ├── Helper
            ├── Model
            ├── Observer
            ├── Setup
            ├── Test
            ├── etc
            ├── i18n
            └── view
    </pre>
  8. Go to the magento root folder (where composer.json is located)

    $ cd $MAGENTO_FOLDER$

  9. Execute Magento setup upgrade

    $ bin/magento setup:upgrade

  10. Clean cache and generated code

    $ bin/magento cache:clean
    
    $ rm -rf var/generation/*

  11. Run magento compiler to generate auto-generated classes

    $ bin/magento setup:di:compile

   (this will take some time ...)

__Test__

  You can check if the module was properly installed testing some features introduced by Sage Pay Suite:
  
  1. Get access to the Magento 2 backoffice.

  2. Menu > Stores > Configuration > SALES > Payment Methods
  You should see Sage Pay Suite on the payment methods list.
  3. Enter your Sage Pay vendorname and Ebizmarts license key on the configuration settings.
  4. Enable the integration of your preference.

[![Build Status](https://circleci.com/gh/ebizmarts/magento2-sage-pay-suite.svg?style=shield&circle-token=9d950c73b76af8868862caf8400c549439838d47)](https://circleci.com/gh/ebizmarts/magento2-sage-pay-suite)


CHANGELOG.md
## [1.3.40] - 2021-07-07
### Fixed
- Module not recovering cart when PI 3D fails.
- Fraud check failing after Opayo update
- The "ResultInterface" class doesn't exist and the namespace must be specified.

## [1.3.39] - 2021-05-26
### Added
- Compatibility with Magento 2.3.7
- Debug Mode setting
- Prevent customer personal data from logging setting
- Show 3rdMan score and score breakdown on order details
### Changed
- Ask customers if want to save the credit card when they already have tokens
### Fixed
- 0.01 difference when you try to invoice PI Defer orders
- DropIn form not appearing after deleting all tokens on checkout
- The service interface name "Ebizmarts\SagePaySuite\Model\Token\Get" is invalid.
- Invoice created successfully in Magento when transaction was aborted
- PI Authorize and Capture orders not being invoiced
- Recover cart message appearing in product page after successful order with PI and 3D

## [1.3.19.1] - 2021-05-12
### Fixed
- PI with 3D redirecting to cart after checkout

## [1.3.19] - 2021-02-02
### Fixed
- Composer installation problem

## [1.3.18] - 2021-02-01
### Changed
- Added token with vault usage on PI.

### Fixed
- PI repeat with 3Dv2
- Recover cart when session is lost
- Fraud not being retrieved for non default sotres in multi-store setup
- Verification result not showing
- Browser Ipv6 error on PI

## [1.3.17] - 2020-11-24
### Fixed
- 3Dv1 not working with Protocol 4.00 for PI
- PI refund problem with Multi-Store sites
- Duplicated Callbacks received for FORM

## [1.3.16] - 2020-10-27
### Changed
- Server cancel payment redirection to checkout shipping method
- Added new Order Details fields names in block

### Fixed
- Fix duplicate 3D callback and duplicate response for threeDSubmit
- CSP Whitelisting file
- Add routes to webRestrctions.xml to avoid payment failures when Magento's EE restrictions is enabled

## [1.3.15] - 2020-07-08
### Changed
- Sage Pay text and logo changed to Opayo

### Fixed
- Adapt 3Dv2 to latest updates
- Duplicated address problem
- 3D, Address, Postcode and CV2 flags not showing up on the order grid
- Recover Cart problem when multiple items with same configurable parent
- Order cancelled when same increment id on different store views
- Duplicated PI Callbacks received cancel the order
- Server not recovering cart when cancel the transaction
- Add form validation in PI WITHOUT Form

## [1.3.14] - 2020-04-13
### Fixed
- Fix PI not working with virtual product and guest checkout
- Problem with js calls not being sequential and causing errors on checkout with PI
- Amount is not an integer when trying to make a refund

## [1.3.13] - 2020-03-18
### Changed
- Store SecurityKey in Database when SyncFromApi
- Enhance cart recovery to avoid orders cancelling when customer goes to checkout/cart
- Round amount when populating amount and currency to avoid 0.01 difference

### Fixed
- Orders cancelled when same IncrementId on different Store Views
- Order not available error with FORM

## [1.3.12] - 2020-02-12
## Added
- Compatibility with Magento 2.3.4
- Show verification results in payment layout at order details

### Changed
- Look transaction by vendorTxCode if not VPSTxId when SyncFromApi

### Fixed
- Problem with basket format when using Sage50
- Error while trying to cancel SERVER Authenticate order

## [1.3.11] - 2019-12-19
### Fixed
- Items being canceled when order take more than 15 minutes
- Guest order being created with "Guest" as customer name
- Pi not loading when there are terms and conditions

## [1.3.10] - 2019-11-26
### Added
- Show Fraud information on order grid (3D, Post Code, Address, CV2)

### Changed
- New PI endpoint

### Fixed
- Order failing if using special characters on order id prefix
- Fraud flag showing no flag when 3rd Man and there's no Fraud Rule

### Security
- Encrypt PI callback URL

## [1.3.9] - 2019-10-28
### Added
- Compatibility with Magento 2.3.3 and 2.3.2-p2
- Setting to open 3D verification in new window for PI

### Changed
- Sanitize Post Code on PI
- Remove spaces from paRes
- Remove "Load secure credit card form" PI button

### Fixed
- Frontend using Default Config values instead of Store values on Frontend
- Multiple 3D responses problem

## [1.3.8] - 2019-10-01
### Added
- PI support for PSD2 and SCA
- Payment Failed Emails impelentation for PI

### Fixed
- Fix DropIn not working with Minify js
- Stop the order for try to being captured if txstateid empty
- 0.00 cost products breaks PayPal
- Fix Multi Currency Authenticate invoice using Base Currency amount

## [1.3.7] - 2019-08-08
### Added
- Setting to set max tokens per customer

### Changed
- Hide Add New Card when reached max tokens
- Change excluding minification strategy

### Fixed
- Label and Checkbox from first token being shown when press add new card
- Send 000 post code when field is left empty for Ireland or Hong Kong (SERVER and FORM)
- PI always sending 000 post code for Ireland and Hong Kong even if the customer entered a post code
- Module breaks Sales -> Order when the payment additional information is serialized
- Multi Currency refunds using Base Currency amount (FORM, SERVER, PayPal)

## [1.3.6] - 2019-06-24
### Added
- SERVER and FORM support for PSD2 and SCA
- PI DropIn compatibility with OneStepCheckout

### Fixed
- Module breaks Sales -> Order
- Server defer orders not being cancelled on SagePay
- Problem with submit payment button PI
- PI always selected as default payment method on the checkout

## [1.3.5] - 2019-05-08
### Added
- Explanation message to order view
- Add waiting for score and test fraud flags
- Add CardHolder Name field to PI without DropIn

### Changed
- Update README.md to use url sagepaysuite.gitlab.ebizmarts.com for composer config.

### Fixed
- PI DropIn MOTO problem with multiple storeviews
- Invoice and Refund problem with multi currency site and base currency
- Basket Sage50 doesn't send space character

### Removed
- PHP restrictions on module for M2.1

Model/Api/Pi.php
<?php
/**
 * Copyright © 2017 ebizmarts. All rights reserved.
 * See LICENSE.txt for license details.
 */

namespace Ebizmarts\SagePaySuite\Model\Api;

use Ebizmarts\SagePaySuite\Helper\Data;
use Ebizmarts\SagePaySuite\Helper\Request;
use Ebizmarts\SagePaySuite\Model\Api\PIRest;
use Ebizmarts\SagePaySuite\Model\Api\Reporting;
use Ebizmarts\SagePaySuite\Model\Config;
use Ebizmarts\SagePaySuite\Model\Logger\Logger;
use Magento\Framework\Phrase;

/**
 * Sage Pay Pi API
 */
class Pi implements PaymentOperations
{
    /** @var \Ebizmarts\SagePaySuite\Model\Api\Reporting */
    private $reportingApi;

    /** @var Data */
    private $suiteHelper;

    /** @var \Ebizmarts\SagePaySuite\Model\Api\PIRest */
    private $piRestApi;

    private $invalidTransStates = [
        8 => 'Transaction CANCELLED by Sage Pay after 15 minutes of inactivity.  This is normally because the customer closed their browser.',
        10 => 'Transaction REJECTED by the Fraud Rules you have in place',
        12 => 'Transaction DECLINED by the bank (NOTAUTHED)',
        11 => 'Transaction ABORTED by the Customer on the Payment Pages',
        13 => 'An ERROR occurred at Sage Pay which cancelled this transaction',
        27 => 'DEFERRED transaction that expired before it was RELEASEd or ABORTed',
        18 => 'Transaction VOIDed by the Vendor',
        19 => 'Successful DEFERRED transaction ABORTED by the Vendor',
        30 => 'The transaction failed',
        32 => 'The transaction was aborted by the customer'
    ];

    /**
     * Pi constructor.
     * @param Data $suiteHelper
     * @param \Ebizmarts\SagePaySuite\Model\Api\PIRest $piRestApi
     * @param \Ebizmarts\SagePaySuite\Model\Api\Reporting $reportingApi
     */
    public function __construct(
        Data $suiteHelper,
        PIRest $piRestApi,
        Reporting $reportingApi
    ) {
        $this->suiteHelper         = $suiteHelper;
        $this->piRestApi           = $piRestApi;
        $this->reportingApi        = $reportingApi;
    }

    public function captureDeferredTransaction($vpsTxId, $amount, \Magento\Sales\Api\Data\OrderInterface $order)
    {
        $result = null;

        $vpsTxId = $this->suiteHelper->clearTransactionId($vpsTxId);
        $transaction = $this->reportingApi->getTransactionDetailsByVpstxid($vpsTxId, $order->getStoreId());

        $this->validateTxStateId($transaction);

        $txStateId = (int)$transaction->txstateid;
        if ($txStateId === PaymentOperations::DEFERRED_AWAITING_RELEASE) {
            $result = $this->piRestApi->release($vpsTxId, $amount);
        } else {
            if ($txStateId === PaymentOperations::SUCCESSFULLY_AUTHORISED) {
                $data                 = [];
                $data['VendorTxCode'] = $this->suiteHelper->generateVendorTxCode('', Config::ACTION_REPEAT_PI);
                $data['Description']  = 'REPEAT deferred transaction from Magento.';
                $data['Currency']     = (string)$transaction->currency;
                $data['Amount']       = $amount * 100;
                $result               = $this->repeatTransaction($vpsTxId, $data, $order, Config::ACTION_REPEAT_PI);
            }
        }

        return $result;
    }

    public function repeatTransaction($vpstxid, $quote_data, \Magento\Sales\Api\Data\OrderInterface $order, $paymentAction = Config::ACTION_REPEAT)
    {
        return $this->piRestApi->repeat(
            $quote_data['VendorTxCode'],
            $vpstxid,
            $quote_data['Currency'],
            $quote_data['Amount'],
            $quote_data['Description']
        );
    }

    public function authorizeTransaction($vpstxid, $amount, \Magento\Sales\Api\Data\OrderInterface $order)
    {
        throw new \Exception("not implented.");
    }

    public function refundTransaction($vpstxid, $amount, \Magento\Sales\Api\Data\OrderInterface $order)
    {
        throw new \Exception("not implented.");
    }

    /**
     * 1 = Transaction failed registration.  Either an INVALID or MALFORMED response was returned
     * 2 = User on Card Selection page
     * 3 = User on the Card Details Entry Page
     * 4 = User on Confirmation Page
     * 5 = Transaction at 3D-Secure Authentication Stage
     * 6 = Transaction sent for Authorisation
     * 7 = Vendor Notified of transaction state at their NotificationURL.  Awaiting response.
     * 8 = Transaction CANCELLED by Sage Pay after 15 minutes of inactivity.  This is normally because the customer closed their browser.
     * 9 = Transaction completed but Vendor systems returned INVALID or ERROR in response to notification POST. Transaction CANCELLED by the Vendor.
     * 10 = Transaction REJECTED by the Fraud Rules you have in place
     * 11 = Transaction ABORTED by the Customer on the Payment Pages
     * 12 = Transaction DECLINED by the bank (NOTAUTHED)
     * 13 = An ERROR occurred at Sage Pay which cancelled this transaction
     * 14 = Successful DEFERRED transaction, awaiting RELEASE
     * 15 = Successful AUTHENTICATED transaction, awaiting AUTHORISE
     * 16 = Successfully authorised transaction
     * 17 = Transaction Timed Out at Authorisation Stage
     * 18 = Transaction VOIDed by the Vendor
     * 19 = Successful DEFERRED transaction ABORTED by the Vendor
     * 20 = Transaction has been timed out by Sage Pay
     * 21 = Successfully REGISTERED transaction, awaiting AUTHORISE
     * 22 = AUTHENTICATED or REGISTERED transaction CANCELLED by the Vendor
     * 23 = Transaction could not be settled with the bank and has been failed by the Sage Pay systems
     * 24 = PayPal Transaction Registered
     * 25 = Token Registered
     * 26 = AUTHENTICATE transaction that can no longer be AUTHORISED against.  It has either expired, or been fully authorised
     * 27 = DEFERRED transaction that expired before it was RELEASEd or ABORTed
     * 28 = Transaction waiting for authorisation
     * 29 = Successfully authorised transaction
     * 30 = The transaction failed
     * 31 = The transaction failed due to invalid or incomplete data
     * 32 = The transaction was aborted by the customer
     * 33 = Transaction timed out at authorisation stage
     * 34 = A remote ERROR occurred at Sage Pay which cancelled this transaction
     * 35 = A local ERROR occurred at Sage Pay which cancelled this transaction
     * 36 = The transaction could not be sent to the bank and has been failed by the Sage Pay systems
     * 37 = The transaction was declined by the bank
     * 38 = User at bank details page
     * 39 = User at Token Details page
     * 40 = TBC
     * 41 = PPro Transaction CANCELLED by Sage
     * @param $transaction
     *
     * @throws ApiException
     */
    private function validateTxStateId($transaction)
    {
        if (property_exists($transaction, "txstateid")) {
            $txstateid = $transaction->txstateid;
            if (isset($this->invalidTransStates[$txstateid])) {
                throw new ApiException(__('Cannot capture deferred transaction, ' . $this->invalidTransStates[$txstateid] . '.'));
            }
        } else {
            throw new ApiException(__('Cannot capture deferred transaction, transaction state is invalid.'));
        }
    }
}
